renku/service/controllers/cache_files_upload.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Renku service cache upload files controller."""
+import os
+import shutil
+from pathlib import Path
+
+import patoolib
+from patoolib.util import PatoolError
+
+from renku.core.errors import RenkuException
+from renku.service.config import CACHE_UPLOADS_PATH, SUPPORTED_ARCHIVES
+from renku.service.controllers.api.abstract import ServiceCtrl
+from renku.service.controllers.api.mixins import ReadOperationMixin
+from renku.service.serializers.cache import FileUploadRequest, FileUploadResponseRPC, extract_file
+from renku.service.views import result_response
+
+
+class UploadFilesCtrl(ServiceCtrl, ReadOperationMixin):
+    """Controller for migrations check endpoint."""
+
+    REQUEST_SERIALIZER = FileUploadRequest()
+    RESPONSE_SERIALIZER = FileUploadResponseRPC()
+
+    def __init__(self, cache, user_data, flask_request):
+        """Construct controller."""
+        self.file = extract_file(flask_request)
+
+        self.response_builder = {
+            "file_name": self.file.filename,
+            "content_type": self.file.content_type,
+            "is_archive": self.file.content_type in SUPPORTED_ARCHIVES,
+        }
+        self.response_builder.update(UploadFilesCtrl.REQUEST_SERIALIZER.load(flask_request.args))
+
+        super(UploadFilesCtrl, self).__init__(cache, user_data, {})
+
+    @property
+    def context(self):
+        """Controller operation context."""
+        return self.response_builder
+
+    def process_file(self):
+        """Process uploaded file."""
+        user_cache_dir = CACHE_UPLOADS_PATH / self.user.user_id
+        user_cache_dir.mkdir(exist_ok=True)
+
+        file_path = user_cache_dir / self.file.filename
```

Comment:
This will never cause a conflict to cause it's controlled through the `override_existing` flag in the request. If you are trying to upload a file that already exists, you will receive an error response describing that, and if you want to override the cached object you just set the `override_existing` to `true` in the request and it will be overridden.

Reply:
I meant if there are concurrent uploads (e.g. a directory upload). I take your word for it that it won't happen since AFAIK we only allow uploading files in the UI.